src/lib/homestore_backend/gc_manager.cpp

```diff
@@ -343,88 +467,390 @@ sisl::sg_list GCManager::pdev_gc_actor::generate_shard_super_blk_sg_list(shard_i
+                                    &data_service, footer_sgs](auto&& results) {
+                            for (auto const& ok : results) {
+                                RELEASE_ASSERT(ok.hasValue(), "we never throw any exception when copying data");
+                                if (!ok.value()) {
+                                    // if any op fails, we drop this gc task.
+                                    return folly::makeFuture< std::error_code >(
+                                        std::make_error_code(std::errc::operation_canceled));
+                                }
+                            }
+
+                            // the shard that triggers emergent gc should be the last shard in the chunk, so it should
+                            // be open and we skip writing the footer for this case.
+                            if (is_emergent && is_last_shard) {
+                                LOGDEBUG(
+                                    "skip writing the footer for move_to_chunk={} shard_id={} for emergent gc task",
+                                    move_to_chunk, shard_id);
+                                return folly::makeFuture< std::error_code >(std::error_code{});
+                            }
+                            homestore::MultiBlkId out_blkids;
+                            return data_service.async_alloc_write(footer_sgs, hints, out_blkids);
+                        })
+                        .thenValue([this, &move_to_chunk, &shard_id, footer_sgs](auto&& err) {
+                            RELEASE_ASSERT(footer_sgs.iovs.size() == 1, "footer_sgs.iovs.size() should be 1, but not!");
+                            iomanager.iobuf_free(reinterpret_cast< uint8_t* >(footer_sgs.iovs[0].iov_base));
+
+                            if (err) {
+                                LOGERROR("Failed to write shard footer for move_to_chunk={} shard_id={}, "
+                                         "err={}",
+                                         move_to_chunk, shard_id, err.value());
+                                return false;
+                            }
+                            return true;
+                        });
+                })
+                .get();
+
+        if (!succeed_copying_shard) {
+            LOGERROR("Failed to copy all blobs from move_from_chunk={} to move_to_chunk={} for shard_id={}",
+                     move_from_chunk, move_to_chunk, shard_id);
+            return false;
+        }
+
+        LOGDEBUG("successfully copy blobs from move_from_chunk={} to move_to_chunk={} for shard_id={}", move_from_chunk,
+                 move_to_chunk, shard_id);
+    }
+
+    LOGDEBUG("all valid blobs are copied from move_from_chunk={} to move_to_chunk={}", move_from_chunk, move_to_chunk);
+
+    // we need to commit_blk for the move_to_chunk to make sure the last offset of append_blk_allocator is updated.
```

Comment:
If we refactor this function a bit
1. lets sync write header via async_alloc_write().get().
2. parallel submit for all data blobs , returns PBA from the async_alloc_write().thenValue().
3. wait all data write finish , get the max PBA
4. (except last shard)  sync write footer.

This function would be more readable as well as we can get the PBA we need to commit.

Reply:
I make this fully aysnc and if any op fails, we fails the whole copy_shard. so no need to check all kinds of the failed cases.
is it more simple to commit to the last blk? I perfer this way.

in the  last shard in emergent chunk, if all the blobs of this shard are written concurrently, we can not make sure which blk is last written. you mean compare the blk one by one?

Reply:
yes a for loop serves the purpose. 
making an async function complicate to 100+ lines then follow a get() is not that "async" . Personally I would prefer just async those needed part(data copy) and leave other part (hearder/footer) sync as I failed to follow the complicate then-value structure.

But anyway , it is your choice.